### DISCUSSION TITLE
GNNs need Sparse Tensors; can we do this?

### DISCUSSION THREAD

**Discussion by santi-mir** [keras-team/keras#20500] [Q&A]
Hi everyone,

**Motivation**

I describe very briefly what I'd like to do (related to OP.), question is at the bottom. 

Here is a [very well cited and simple paper](https://arxiv.org/abs/1609.02907) (40K) for convolutions on graphs (I'd like to use it for chemistry.)

**Torch Geometric**

It would be easy with [Torch Geometric (**PyG**)](https://github.com/pyg-team/pytorch_geometric):

> PyG (PyTorch Geometric) is a library built upon [PyTorch](https://pytorch.org/) to easily write and train Graph Neural Networks (GNNs) for a wide range of applications related to structured data

_I want to use Keras for multi backend._

The formula in the paper seems trivial to implement in a layer:

![Screenshot from 2024-11-15 16-11-03](https://github.com/user-attachments/assets/11e32b8f-82e3-4560-a829-7d544b6ef8a5)

but they talk about linear increase of complexity with the number of edges, surely this is due to sparse matrices.  

![Screenshot from 2024-11-15 16-02-44](https://github.com/user-attachments/assets/07342f19-f224-4ca6-a225-f59a37561623)

**Question 0**: Yet, it seems [Sparse Tensors support](https://github.com/keras-team/keras/issues/18420) still on queue? 

**Question 1**: We need to re-implement the graph-like layers to get multi-backend, right ? Otherwise one would just PyG. 

**Question 2:** Can one use such sparse representations in pure Keras, chaging then the backend ? I can't find a tutorial or ref, [apart from TF itself here.](https://www.tensorflow.org/guide/sparse_tensor)

**Comment by Kxrma47** [comment]
You do not necessarily need a sparse-matrix type to keep graph convolution at `O(E)`. A portable representation is an edge list plus node features:

```text
node_features: [num_nodes, channels]
edge_index:    [2, num_edges]   # source, destination
```

A Keras layer can gather source-node features, compute messages, and aggregate them with `keras.ops.segment_sum` (or the corresponding scatter operation). That touches each edge once and avoids materializing a dense `N x N` adjacency matrix.

Conceptually:

```python
src, dst = edge_index[0], edge_index[1]
messages = keras.ops.take(node_features, src, axis=0)
aggregated = keras.ops.segment_sum(
    messages, dst, num_segments=keras.ops.shape(node_features)[0]
)
```

Then add normalization, self-loops, and a learned projection according to the exact GCN formulation. For batches of graphs, concatenate their nodes and offset each graph's edge indices, while retaining a graph-id vector for pooling.

There is no single sparse tensor abstraction whose storage, autograd, and supported kernels are identical across TensorFlow, JAX, and PyTorch. Using a backend-native sparse tensor therefore gives up some portability. Edge-list message passing built from `keras.ops` is usually the better multi-backend starting point; backend-specific libraries such as PyG remain preferable when you need their optimized operators, samplers, and large graph ecosystem.
